packages/game-passes/contracts/GamePasses.sol

```diff
@@ -0,0 +1,1161 @@
+     *      - Token is not configured
+     *      - Signature is invalid or expired
+     *      - Max supply would be exceeded
+     *      - Max per wallet would be exceeded
+     *      - Payment transfer fails
+     */
+    function mint(
+        address from,
+        uint256 tokenId,
+        uint256 amount,
+        uint256 price,
+        uint256 deadline,
+        bytes memory signature
+    ) external whenNotPaused {
+        address caller = _msgSender();
+
+        if (from != caller) {
+            revert InvalidSender(from, caller);
+        }
+
+        TokenConfig storage config = tokenConfigs[tokenId];
+
+        if (!config.isConfigured) {
+            revert TokenNotConfigured(tokenId);
+        }
+        MintRequest memory request = MintRequest({
+            caller: caller,
+            tokenId: tokenId,
+            amount: amount,
+            price: price,
+            deadline: deadline,
+            nonce: nonces[caller]++
+        });
+
+        verifySignature(request, signature);
+        _checkMaxPerWallet(tokenId, caller, amount);
+        _checkMaxSupply(tokenId, amount);
+
+        // Update minted amount for wallet
+        config.mintedPerWallet[caller] += amount;
+
+        // Transfer payment
+        address treasury = config.treasuryWallet != address(0) ? config.treasuryWallet : defaultTreasuryWallet;
+        SafeERC20.safeTransferFrom(IERC20(paymentToken), _msgSender(), treasury, price);
+
+        // Mint tokens
+        _mint(caller, tokenId, amount, "");
+    }
+
+    /**
+     * @notice Batch mints multiple tokens with valid EIP-712 signatures in a single transaction
+     * @param from Address that will receive the tokens (must be same as msg.sender)
+     * @param tokenIds Array of token IDs to mint
+     * @param amounts Array of amounts to mint for each token ID
+     * @param prices Array of prices to pay for each mint operation
+     * @param deadlines Array of timestamps after which each signature becomes invalid
+     * @param signatures Array of EIP-712 signatures from authorized signers
+     * @dev Processes multiple mint operations in batch, verifying each signature
+     * @dev All array parameters must be the same length
+     * @dev Updates per-wallet minting counts and transfers payments to appropriate treasuries
+     * @dev Reverts if:
+     *      - Contract is paused
+     *      - Array lengths don't match
+     *      - Any token is not configured
+     *      - Any signature is invalid or expired
+     *      - Any max supply would be exceeded
+     *      - Any max per wallet would be exceeded
+     *      - Any payment transfer fails
+     */
+    function batchMint(
+        address from,
+        uint256[] calldata tokenIds,
+        uint256[] calldata amounts,
+        uint256[] calldata prices,
+        uint256[] calldata deadlines,
+        bytes[] calldata signatures
+    ) external whenNotPaused {
+        address caller = _msgSender();
```

Comment:
maybe to is better than from ?
why don't you use msg.sender (or _msgSender) directly instead of passing an extra argument ?

Reply:
this is for approveAndCall where the sender must be the first argument

its only for user facing functions that may be called through SAND contract

Reply:
To use approveAndCall you must accept the calls when the caller is the Sand contract, but in the code it seems that you reject them or maybe I'm missing something?

Reply:
What do you mean I reject them? How?

I want to get past this check in the approveAndCall

```solidity
require(BytesUtil.doFirstParamEqualsAddress(data, _msgSender()), "FIRST_PARAM_NOT_SENDER");
```

Let me know if I am doing something wrong here.

Reply:
I think that when you call approveAndCall the `caller == msg.sender == _msgSender()` is the `sandContractAddress`  which is different from `from` and then the contract reverts with `InvalidSender` (if I'm not missing something)

Reply:
Thanks @adjisb, I got the issue finally!

Please check the fix: https://github.com/thesandboxgame/sandbox-smart-contracts/pull/1619/commits/fe367759bbfee4bdd065d6786ac2231a00b8b9fc